### PR TITLE
[Feature] Add organizer to FS

### PR DESCRIPTION
**Pull request by rayrayraykk**
# Organizer

Organizer is implemented by [Celery](https://docs.celeryq.dev/en/latest/) and [Redis](https://redis.io/) as Broker, so you should install it before using.

## Installation

```bash
# For Server
pip install -e .[org]
docker run -d -p 6379:6379 redis

# For Client
pip install -e .[org]
# Not full login ssh, will conduct `source ~/.bashrc` first.
# ** If not running interactively, `source ~/.bashrc` might fail **
# ** due to: `[ -z "$PS1" ] && return`, please comment this line **
```

## RUN

```bash
# For Server
cd federatedscope/organizer
celery -A server worker --loglevel=info

## For multi-worker
# celery multi start w1 -A server -l info
# celery multi start w2 -A server -l info
# ...

# For Client
# Modify `server_ip` in federatedscope/organizer/cfg_client.py
python federatedscope/organizer/client.py
help
```

### CODE REVIEW

File: federatedscope/organizer/server.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            'cfg': args2yaml(args),
+            'psw': psw,
+        }
+        if room_id in room.keys():
+            raise ValueError
+        else:
+            room[room_id] = meta_info
+        self._save('room', room)
+
+        # Launch FS
```
Comment (rayrayraykk) [inline federatedscope/organizer/server.py:80]: I will add `exception_handler` here in the next PR, when launching FS failed.

File: federatedscope/core/configs/config.py [hunk 2]
```diff
@@ -238,37 +239,42 @@ def freeze(self, inform=True):
...
-                        "cfg.wandb.use=True but not install the wandb package")
-                    exit()
-
-        if inform:
+        if save:  # save the final cfg
+            Path(self.outdir).mkdir(parents=True, exist_ok=True)
+            with open(os.path.join(self.outdir, "config.yaml"),
+                      'w') as outfile:
+                from contextlib import redirect_stdout
+                with redirect_stdout(outfile):
```
Comment (joneswong) [inline federatedscope/core/configs/config.py:247]: @yxdyc What's the reason to write files via redirecting the standard output stream?

File: federatedscope/core/configs/config.py [hunk 3]
```diff
@@ -238,37 +239,42 @@ def freeze(self, inform=True):
...
+                        import yaml
+                        cfg_yaml = yaml.safe_load(tmp_cfg.dump())
+                        wandb.config.update(cfg_yaml, allow_val_change=True)
+                    except ImportError:
+                        logger.error(
+                            "cfg.wandb.use=True but not install the wandb "
+                            "package")
+                        exit()
+
+        if inform and save:
```
Comment (joneswong) [inline federatedscope/core/configs/config.py:264]: why not just put this if branch in the above one

File: federatedscope/core/configs/config.py (hunk 2 above)
Comment (yxdyc) [reply federatedscope/core/configs/config.py:247]: No special considerations, just personal convention, and I tested that the way it has been written makes the later `yaml.safe_load()` work correctly.  Perhaps changing the code here into sth. with `yaml.safe_dump()` would make it look more elegant.

**Review by joneswong — APPROVED** [review]
LGTM

File: federatedscope/core/configs/config.py (hunk 3 above)
Comment (rayrayraykk) [reply federatedscope/core/configs/config.py]: Update accordingly.

**Pull request merged**